### ISSUE TITLE
I output gradient values during training and observed that gradient explosion has always existed, but as long as nan values do not appear, it does not seem to affect the training effect. However, more and more gradient disappearance problems have made it impossible for my network to learn useful information.

### ISSUE THREAD

**Issue by yueyangwen** [Pointcept/PointTransformerV3#54]
Hello, I trained on my two-category semantic segmentation task. I found that the first few pieces of data trained well, but the performance will be very poor later. I output the gradient information to view it. I found that as the network trains, More and more layers are experiencing the vanishing gradient problem, but I cannot add a bn layer to the network. Do you have any suggestions?

**Comment by Gofinge** [comment]
Maybe you can try to add a normalization in QKV project just as PTv2?

**Comment by JamesMcCullochDickens** [comment]
I also notice gradient explosions giving me nans and infs, I suspect some additional normalization might help, but I'm not sure where and for which layer. For now I am using gradient clipping at a value of 1.0 (torch.nn.utils.clip_grad_norm_(model.parameters(), grad_clip_val))

It seems to be related more to the 3D sparse convolutions for my use case rather than the attention stuff. I'm curious if you have tried using regular batch norm here instead of layer norm? In my mind the layer norm makes more sense, but I do see examples in spconv using 3d convolutions followed by batch norm. I'll give it a shot and report back at some point.

**Comment by SpeedyGonzales949** [comment]
> Maybe you can try to add a normalization in QKV project just as PTv2?

do you mean at the start or at the end of the attention layer ?

**Comment by Gofinge** [comment]
> I also notice gradient explosions giving me nans and infs, I suspect some additional normalization might help, but I'm not sure where and for which layer. For now I am using gradient clipping at a value of 1.0 (torch.nn.utils.clip_grad_norm_(model.parameters(), grad_clip_val))
> 
> It seems to be related more to the 3D sparse convolutions for my use case rather than the attention stuff. I'm curious if you have tried using regular batch norm here instead of layer norm? In my mind the layer norm makes more sense, but I do see examples in spconv using 3d convolutions followed by batch norm. I'll give it a shot and report back at some point.

We used to apply BN in v1 and v2; (Check our ablation in paper) we want to enable LN as it is independent of batch size. Nice try, waiting for your response.

**Comment by AxelCole** [comment]
> I also notice gradient explosions giving me nans and infs, I suspect some additional normalization might help, but I'm not sure where and for which layer. For now I am using gradient clipping at a value of 1.0 (torch.nn.utils.clip_grad_norm_(model.parameters(), grad_clip_val))
> 
> It seems to be related more to the 3D sparse convolutions for my use case rather than the attention stuff. I'm curious if you have tried using regular batch norm here instead of layer norm? In my mind the layer norm makes more sense, but I do see examples in spconv using 3d convolutions followed by batch norm. I'll give it a shot and report back at some point.

Any update on this ? 
Has anyone found the origin of the gradient explosion and any way to fix it without gradient clipping ?

**Comment by JamesMcCullochDickens** [comment]
@AxelCole 

If you look at the spconv library, they have 1D batch norm layers on the feature channels of the points, try replacing the 3D submanifold convolution layers with a sparse sequential using 3d submanifold convolution, batch norm and relu, might do the trick and add some stability.

**Comment by Gofinge** [comment]
Clip gradient is very easy to apply (I will update it in the next version), like below:

(pointcept.engine.train.py)
```
def run_step(self):
        input_dict = self.comm_info["input_dict"]
        for key in input_dict.keys():
            if isinstance(input_dict[key], torch.Tensor):
                input_dict[key] = input_dict[key].cuda(non_blocking=True)
        with torch.cuda.amp.autocast(enabled=self.cfg.enable_amp):
            output_dict = self.model(input_dict)
            loss = output_dict["loss"]
        self.optimizer.zero_grad()
        if self.cfg.enable_amp:
            self.scaler.scale(loss).backward()
            self.scaler.unscale_(self.optimizer)
            if self.cfg.clip_grad is not None:
                torch.nn.utils.clip_grad_norm_(
                    self.model.parameters(), self.cfg.clip_grad
                )
            self.scaler.step(self.optimizer)

            # When enable amp, optimizer.step call are skipped if the loss scaling factor is too large.
            # Fix torch warning scheduler step before optimizer step.
            scaler = self.scaler.get_scale()
            self.scaler.update()
            if scaler <= self.scaler.get_scale():
                self.scheduler.step()
        else:
            loss.backward()
            if self.cfg.clip_grad is not None:
                torch.nn.utils.clip_grad_norm_(
                    self.model.parameters(), self.cfg.clip_grad
                )
            self.optimizer.step()
            self.scheduler.step()

        if self.cfg.empty_cache:
            torch.cuda.empty_cache()
        self.comm_info["model_output_dict"] = output_dict
```

<img width="915" alt="image" src="https://github.com/user-attachments/assets/9b960c49-f7f4-410b-8fa4-0c3a8b6df51c">
